riscv/devices.cc

```diff
@@ -10,53 +10,74 @@ mmio_device_map_t& mmio_device_map()
-  // container to sort the keys and provide ordered
-  // iteration over this sort, which it does. (python's
-  // SortedDict is a good analogy)
+  // Allow empty devices by omitting them
+  auto size = dev->size();
+  if (size == 0)
+    return;
+
+  // Reject devices that overflow address size
+  if (addr + size - 1 < addr) {
+    fprintf(stderr, "device at [%" PRIx64 ", %" PRIx64 ") overflows address size\n",
+            addr, addr + size);
+    abort();
+  }
+
+  // Reject devices that overlap other devices
+  if (auto it = devices.upper_bound(addr);
+      (it != devices.end() && addr + size - 1 >= it->first) ||
+      (it != devices.begin() && (it--, it->first + it->second->size() - 1 >= addr))) {
+    fprintf(stderr, "devices at [%" PRIx64 ", %" PRIx64 ") and [%" PRIx64 ", %" PRIx64 ") overlap\n",
+            it->first, it->first + it->second->size(), addr, addr + size);
+    abort();
+  }
+
   devices[addr] = dev;
 }
 
 bool bus_t::load(reg_t addr, size_t len, uint8_t* bytes)
 {
-  // Find the device with the base address closest to but
-  // less than addr (price-is-right search)
-  auto it = devices.upper_bound(addr);
-  if (devices.empty() || it == devices.begin()) {
-    // Either the bus is empty, or there weren't 
-    // any items with a base address <= addr
-    return false;
-  }
-  // Found at least one item with base address <= addr
-  // The iterator points to the device after this, so
-  // go back by one item.
-  it--;
-  return it->second->load(addr - it->first, len, bytes);
+  if (auto [base, dev] = find_device(addr, len); dev)
+    return dev->load(addr - base, len, bytes);
+  return false;
 }
 
 bool bus_t::store(reg_t addr, size_t len, const uint8_t* bytes)
 {
-  // See comments in bus_t::load
-  auto it = devices.upper_bound(addr);
-  if (devices.empty() || it == devices.begin()) {
-    return false;
-  }
-  it--;
-  return it->second->store(addr - it->first, len, bytes);
+  if (auto [base, dev] = find_device(addr, len); dev)
+    return dev->store(addr - base, len, bytes);
+  return false;
+}
+
+reg_t bus_t::size()
+{
+  if (auto last = devices.rbegin(); last != devices.rend())
+    return last->first + last->second->size();
+  return 0;
 }
 
 std::pair<reg_t, abstract_device_t*> bus_t::find_device(reg_t addr)
 {
-  // See comments in bus_t::load
+  // Obtain iterator to device immediately after the one that might match
   auto it = devices.upper_bound(addr);
   if (devices.empty() || it == devices.begin()) {
+    // No devices with base address <= addr
     return std::make_pair((reg_t)0, (abstract_device_t*)NULL);
   }
+
+  // Rewind to device that will match if its size suffices
   it--;
+
   return std::make_pair(it->first, it->second);
 }
 
+std::pair<reg_t, abstract_device_t*> bus_t::find_device(reg_t addr, size_t len)
+{
+  if (auto [base, dev] = find_device(addr); addr - base + len - 1 < dev->size())
+    return std::make_pair(base, dev);
+
+  return std::make_pair(0, nullptr);
```

Comment:
`find_device(addr)` can return null device pointer, so I propose we add this check first

```
  auto [base, dev] = find_device(addr);
  if (dev && addr - base + len - 1 < dev->size())
    return std::make_pair(base, dev);

  return std::make_pair(0, nullptr);
  ```

Reply:
Oops, good catch.